tree/dataframe/inc/ROOT/RDF/RDatasetSpec.hxx

```diff
@@ -0,0 +1,124 @@
+      Long64_t fStartEntry{0};
+      Long64_t fEndEntry{std::numeric_limits<Long64_t>::max()};
+      REntryRange() {}
+      REntryRange(Long64_t endEntry) : fEndEntry(endEntry) {}
+      REntryRange(Long64_t startEntry, Long64_t endEntry)
+         : fStartEntry(startEntry),
+           fEndEntry(endEntry >= startEntry
+                        ? endEntry
+                        : throw std::logic_error("The starting entry cannot be larger than the ending entry in the "
+                                                 "creation of a dataset specification."))
+      {
+      }
+   };
+
+   /**
+    * A list of names of trees.
+    * This list should go in lockstep with fFileNameGlobs, only in case this dataset is a TChain where each file
+    * contains its own tree with a different name from the global name of the dataset.
+    */
+   std::vector<std::string> fTreeNames{};
+
+   /**
+    * A list of file names.
+    * They can contain the globbing characters supported by TChain. See TChain::Add for more information.
+    */
+   std::vector<std::string> fFileNameGlobs{};
+
+   Long64_t fStartEntry{}; ///< The entry where the dataset processing should start (inclusive).
+   Long64_t fEndEntry{};   ///< The entry where the dataset processing should end (exclusive).
+
+   ROOT::Internal::TreeUtils::RFriendInfo fFriendInfo{}; ///< List of friends
+
+   RDatasetSpec(const std::string &treeName, const std::string &fileNameGlob, const REntryRange &entryRange = {})
+      : fTreeNames(std::vector<std::string>{std::move(treeName)}),
+        fFileNameGlobs(std::vector<std::string>{std::move(fileNameGlob)}), fStartEntry(entryRange.fStartEntry),
+        fEndEntry(entryRange.fEndEntry)
+   {
+   }
+
+   RDatasetSpec(const std::string &treeName, const std::vector<std::string> &fileNameGlobs,
+                const REntryRange &entryRange = {})
+      : fTreeNames(std::vector<std::string>{std::move(treeName)}), fFileNameGlobs(std::move(fileNameGlobs)),
+        fStartEntry(entryRange.fStartEntry), fEndEntry(entryRange.fEndEntry)
+   {
+   }
+
+   RDatasetSpec(const std::vector<std::pair<std::string, std::string>> &treeAndFileNameGlobs,
+                const REntryRange &entryRange = {})
+      : fStartEntry(entryRange.fStartEntry), fEndEntry(entryRange.fEndEntry)
+   {
+      fTreeNames.reserve(treeAndFileNameGlobs.size());
+      fFileNameGlobs.reserve(treeAndFileNameGlobs.size());
+      for (auto &p : treeAndFileNameGlobs) {
+         fTreeNames.emplace_back(std::move(p.first));
+         fFileNameGlobs.emplace_back(std::move(p.second));
+      }
+   }
+
+   void AddFriend(const std::string &treeName, const std::string &fileNameGlob, const std::string &alias = "")
+   {
+      fFriendInfo.fFriendNames.emplace_back(std::make_pair(std::move(treeName), std::move(alias)));
+      fFriendInfo.fFriendFileNames.emplace_back(std::vector<std::string>{std::move(fileNameGlob)});
+      fFriendInfo.fFriendChainSubNames.emplace_back(); // this is a tree
```

Comment:
well, we don't want to misalign the friends!

Reply:
I am asking whether we could leave `fFriendChainSubNames` empty instead of filling it with N empty strings -- downstream code would not misalign things if it handles things correctly.

but it could be that it's simpler to allocate N empty strings, I'm not sure.

Reply:
For consistency with what we already have in TProcessorMT, it is easier to keep N strings. After resolving the other comments, I will take a look if it could be changed there as well.

Reply:
maybe @vepadulano already has the answer?